From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Caboose <caboose@invalid>
Date: Sat, 31 Oct 2026 00:00:00 +0000
Subject: [PATCH 14/15] Defer the results of buffer and image memory binds

With guest-assigned handles and a single encoder, vkBindBufferMemory and
vkBindImageMemory (and their 2/2KHR forms) no longer wait for their result.
Their replies are drained in order like other deferred replies; because the
application can no longer observe a failure, any failure is device loss.
---
 .../vulkan_enc/autogenerated/VkEncoder.cpp    | 237 +++++++++++++++---
 1 file changed, 205 insertions(+), 32 deletions(-)

diff --git a/src/gfxstream/guest/vulkan_enc/autogenerated/VkEncoder.cpp b/src/gfxstream/guest/vulkan_enc/autogenerated/VkEncoder.cpp
index cfc4f86..1695734 100644
--- a/src/gfxstream/guest/vulkan_enc/autogenerated/VkEncoder.cpp
+++ b/src/gfxstream/guest/vulkan_enc/autogenerated/VkEncoder.cpp
@@ -2179,18 +2179,29 @@ VkResult VkEncoder::vkAllocateMemory(VkDevice device, const VkMemoryAllocateInfo
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = false && pMemory;
+    if (cabooseGuestHandle) *pMemory = (VkDeviceMemory)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pMemory));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkDeviceMemory(&cgen_var_3, (VkDeviceMemory*)pMemory,
                                                            1);
     stream->unsetHandleMapping();
     VkResult vkAllocateMemory_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkAllocateMemory_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkAllocateMemory);
+    } else {
+        caboose_read_profile::read(stream, &vkAllocateMemory_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -2801,7 +2812,12 @@ VkResult VkEncoder::vkBindBufferMemory(VkDevice device, VkBuffer buffer, VkDevic
     memcpy(*streamPtrPtr, (VkDeviceSize*)&local_memoryOffset, sizeof(VkDeviceSize));
     *streamPtrPtr += sizeof(VkDeviceSize);
     VkResult vkBindBufferMemory_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindBufferMemory_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindBufferMemory);
+    } else {
+        caboose_read_profile::read(stream, &vkBindBufferMemory_VkResult_return, sizeof(VkResult));
+    }
     if (vkBindBufferMemory_VkResult_return == VK_SUCCESS) {
         sResourceTracker->cabooseRecordBufferMemoryBind(buffer, memory, memoryOffset);
     }
@@ -2878,7 +2894,12 @@ VkResult VkEncoder::vkBindImageMemory(VkDevice device, VkImage image, VkDeviceMe
     memcpy(*streamPtrPtr, (VkDeviceSize*)&local_memoryOffset, sizeof(VkDeviceSize));
     *streamPtrPtr += sizeof(VkDeviceSize);
     VkResult vkBindImageMemory_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindImageMemory_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindImageMemory);
+    } else {
+        caboose_read_profile::read(stream, &vkBindImageMemory_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -3462,17 +3483,28 @@ VkResult VkEncoder::vkCreateFence(VkDevice device, const VkFenceCreateInfo* pCre
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = false && pFence;
+    if (cabooseGuestHandle) *pFence = (VkFence)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pFence));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkFence(&cgen_var_3, (VkFence*)pFence, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateFence_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateFence_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateFence);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateFence_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -3876,17 +3908,28 @@ VkResult VkEncoder::vkCreateSemaphore(VkDevice device, const VkSemaphoreCreateIn
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = false && pSemaphore;
+    if (cabooseGuestHandle) *pSemaphore = (VkSemaphore)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pSemaphore));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkSemaphore(&cgen_var_3, (VkSemaphore*)pSemaphore, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateSemaphore_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateSemaphore_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateSemaphore);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateSemaphore_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -4992,18 +5035,29 @@ VkResult VkEncoder::vkCreateCommandPool(VkDevice device, const VkCommandPoolCrea
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pCommandPool;
+    if (cabooseGuestHandle) *pCommandPool = (VkCommandPool)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pCommandPool));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkCommandPool(&cgen_var_3, (VkCommandPool*)pCommandPool,
                                                           1);
     stream->unsetHandleMapping();
     VkResult vkCreateCommandPool_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateCommandPool_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateCommandPool);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateCommandPool_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -7632,18 +7686,29 @@ VkResult VkEncoder::vkCreateShaderModule(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pShaderModule;
+    if (cabooseGuestHandle) *pShaderModule = (VkShaderModule)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pShaderModule));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkShaderModule(&cgen_var_3,
                                                            (VkShaderModule*)pShaderModule, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateShaderModule_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateShaderModule_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateShaderModule);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateShaderModule_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -7815,18 +7880,29 @@ VkResult VkEncoder::vkCreatePipelineCache(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pPipelineCache;
+    if (cabooseGuestHandle) *pPipelineCache = (VkPipelineCache)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pPipelineCache));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkPipelineCache(&cgen_var_3,
                                                             (VkPipelineCache*)pPipelineCache, 1);
     stream->unsetHandleMapping();
     VkResult vkCreatePipelineCache_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreatePipelineCache_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreatePipelineCache);
+    } else {
+        caboose_read_profile::read(stream, &vkCreatePipelineCache_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -8403,18 +8479,29 @@ VkResult VkEncoder::vkCreatePipelineLayout(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pPipelineLayout;
+    if (cabooseGuestHandle) *pPipelineLayout = (VkPipelineLayout)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pPipelineLayout));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkPipelineLayout(&cgen_var_3,
                                                              (VkPipelineLayout*)pPipelineLayout, 1);
     stream->unsetHandleMapping();
     VkResult vkCreatePipelineLayout_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreatePipelineLayout_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreatePipelineLayout);
+    } else {
+        caboose_read_profile::read(stream, &vkCreatePipelineLayout_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -8583,17 +8670,28 @@ VkResult VkEncoder::vkCreateSampler(VkDevice device, const VkSamplerCreateInfo*
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pSampler;
+    if (cabooseGuestHandle) *pSampler = (VkSampler)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pSampler));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkSampler(&cgen_var_3, (VkSampler*)pSampler, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateSampler_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateSampler_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateSampler);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateSampler_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -8767,18 +8865,29 @@ VkResult VkEncoder::vkCreateDescriptorSetLayout(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pSetLayout;
+    if (cabooseGuestHandle) *pSetLayout = (VkDescriptorSetLayout)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pSetLayout));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkDescriptorSetLayout(
         &cgen_var_3, (VkDescriptorSetLayout*)pSetLayout, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateDescriptorSetLayout_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateDescriptorSetLayout_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateDescriptorSetLayout);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateDescriptorSetLayout_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -8954,18 +9063,29 @@ VkResult VkEncoder::vkCreateDescriptorPool(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pDescriptorPool;
+    if (cabooseGuestHandle) *pDescriptorPool = (VkDescriptorPool)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pDescriptorPool));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkDescriptorPool(&cgen_var_3,
                                                              (VkDescriptorPool*)pDescriptorPool, 1);
     stream->unsetHandleMapping();
     VkResult vkCreateDescriptorPool_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateDescriptorPool_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateDescriptorPool);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateDescriptorPool_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -10604,18 +10724,29 @@ VkResult VkEncoder::vkCreateRenderPass(VkDevice device, const VkRenderPassCreate
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pRenderPass;
+    if (cabooseGuestHandle) *pRenderPass = (VkRenderPass)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pRenderPass));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkRenderPass(&cgen_var_3, (VkRenderPass*)pRenderPass,
                                                          1);
     stream->unsetHandleMapping();
     VkResult vkCreateRenderPass_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateRenderPass_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateRenderPass);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateRenderPass_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -12439,7 +12570,12 @@ VkResult VkEncoder::vkBindBufferMemory2(VkDevice device, uint32_t bindInfoCount,
                                                streamPtrPtr);
     }
     VkResult vkBindBufferMemory2_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindBufferMemory2_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindBufferMemory2);
+    } else {
+        caboose_read_profile::read(stream, &vkBindBufferMemory2_VkResult_return, sizeof(VkResult));
+    }
     if (vkBindBufferMemory2_VkResult_return == VK_SUCCESS && pBindInfos) {
         for (uint32_t i = 0; i < bindInfoCount; ++i) {
             sResourceTracker->cabooseRecordBufferMemoryBind(pBindInfos[i].buffer, pBindInfos[i].memory, pBindInfos[i].memoryOffset);
@@ -12524,7 +12660,12 @@ VkResult VkEncoder::vkBindImageMemory2(VkDevice device, uint32_t bindInfoCount,
                                               streamPtrPtr);
     }
     VkResult vkBindImageMemory2_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindImageMemory2_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindImageMemory2);
+    } else {
+        caboose_read_profile::read(stream, &vkBindImageMemory2_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -15367,18 +15508,29 @@ VkResult VkEncoder::vkCreateRenderPass2(VkDevice device, const VkRenderPassCreat
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pRenderPass;
+    if (cabooseGuestHandle) *pRenderPass = (VkRenderPass)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pRenderPass));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkRenderPass(&cgen_var_3, (VkRenderPass*)pRenderPass,
                                                          1);
     stream->unsetHandleMapping();
     VkResult vkCreateRenderPass2_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateRenderPass2_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateRenderPass2);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateRenderPass2_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -21112,18 +21264,29 @@ VkResult VkEncoder::vkCreateRenderPass2KHR(VkDevice device,
     }
     /* is handle, possibly out */;
     uint64_t cgen_var_2;
+    const bool cabooseGuestHandle = caboose_guest_handles::enabled() && pRenderPass;
+    if (cabooseGuestHandle) *pRenderPass = (VkRenderPass)(uintptr_t)caboose_guest_handles::next();
     *&cgen_var_2 = (uint64_t)((*pRenderPass));
     memcpy(*streamPtrPtr, (uint64_t*)&cgen_var_2, 8);
     *streamPtrPtr += 8;
     /* is handle, possibly out */;
     stream->setHandleMapping(sResourceTracker->createMapping());
     uint64_t cgen_var_3;
-    caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    if (cabooseGuestHandle) {
+        cgen_var_3 = cgen_var_2;
+    } else {
+        caboose_read_profile::read(stream, (uint64_t*)&cgen_var_3, 8);
+    }
     stream->handleMapping()->mapHandles_u64_VkRenderPass(&cgen_var_3, (VkRenderPass*)pRenderPass,
                                                          1);
     stream->unsetHandleMapping();
     VkResult vkCreateRenderPass2KHR_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkCreateRenderPass2KHR_VkResult_return, sizeof(VkResult));
+    if (cabooseGuestHandle) {
+        stream->flush();
+        caboose::memory::deferredReplies.addHandleCreate(this, cgen_var_2, packetSize_vkCreateRenderPass2KHR);
+    } else {
+        caboose_read_profile::read(stream, &vkCreateRenderPass2KHR_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
@@ -22128,7 +22291,12 @@ VkResult VkEncoder::vkBindBufferMemory2KHR(VkDevice device, uint32_t bindInfoCou
                                                streamPtrPtr);
     }
     VkResult vkBindBufferMemory2KHR_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindBufferMemory2KHR_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindBufferMemory2KHR);
+    } else {
+        caboose_read_profile::read(stream, &vkBindBufferMemory2KHR_VkResult_return, sizeof(VkResult));
+    }
     if (vkBindBufferMemory2KHR_VkResult_return == VK_SUCCESS && pBindInfos) {
         for (uint32_t i = 0; i < bindInfoCount; ++i) {
             sResourceTracker->cabooseRecordBufferMemoryBind(pBindInfos[i].buffer, pBindInfos[i].memory, pBindInfos[i].memoryOffset);
@@ -22212,7 +22380,12 @@ VkResult VkEncoder::vkBindImageMemory2KHR(VkDevice device, uint32_t bindInfoCoun
                                               streamPtrPtr);
     }
     VkResult vkBindImageMemory2KHR_VkResult_return = (VkResult)0;
-    caboose_read_profile::read(stream, &vkBindImageMemory2KHR_VkResult_return, sizeof(VkResult));
+        if (caboose_guest_handles::enabled() && caboose::memory::deferredReplies.singleEncoder(this)) {
+        stream->flush();
+        caboose::memory::deferredReplies.addCheckedResult(this, packetSize_vkBindImageMemory2KHR);
+    } else {
+        caboose_read_profile::read(stream, &vkBindImageMemory2KHR_VkResult_return, sizeof(VkResult));
+    }
     ++encodeCount;
     if (0 == encodeCount % POOL_CLEAR_INTERVAL) {
         pool->freeAll();
